src/snapshot/src/lib.rs

```diff
@@ -0,0 +1,813 @@
+
+        if computed_checksum != stored_checksum {
+            return Err(Error::Crc64(computed_checksum));
+        }
+
+        Ok(snapshot)
+    }
+
+    /// Saves a snapshot and include a CRC64 checksum.
+    pub fn save_with_crc64<T>(&mut self, writer: &mut T) -> Result<()>
+    where
+        T: std::io::Write,
+    {
+        let mut crc_writer = CRC64Writer::new(writer);
+        self.save(&mut crc_writer)?;
+
+        let checksum = crc_writer.checksum();
+        checksum.serialize(&mut crc_writer, &Self::format_version_map(), 0)?;
+        Ok(())
+    }
+
+    /// Save a snapshot.
+    pub fn save<T>(&mut self, mut writer: &mut T) -> Result<()>
+    where
+        T: std::io::Write,
+    {
+        self.hdr = SnapshotHdr {
+            data_version: self.target_version,
+            section_count: self.sections.len() as u16,
+        };
+
+        let format_version_map = Self::format_version_map();
+        let magic_id = build_magic_id(format_version_map.latest_version());
+
+        // Serialize magic id using the format version map.
+        magic_id.serialize(&mut writer, &format_version_map, 0 /* unused */)?;
+        // Serialize header using the format version map.
+        self.hdr.serialize(
+            &mut writer,
+            &format_version_map,
+            format_version_map.latest_version(),
+        )?;
+
+        // Serialize all the sections.
+        for section in self.sections.values() {
+            // The sections are already serialized.
+            section.serialize(
+                &mut writer,
+                &format_version_map,
+                format_version_map.latest_version(),
+            )?;
+        }
+        writer
+            .flush()
+            .map_err(|ref err| Error::Io(err.raw_os_error().unwrap_or(0)))?;
+
+        Ok(())
+    }
+
+    /// Attempts to find and reads a section (deserialize/translate) from a snapshot.
+    pub fn read_section<T>(&mut self, name: &str) -> Result<T>
+    where
+        T: Versionize,
+    {
+        if let Some(section) = self.sections.get_mut(name) {
+            Ok(T::deserialize(
+                &mut section.data.as_mut_slice().as_ref(),
+                &self.version_map,
+                self.hdr.data_version,
+            )?)
+        } else {
+            Err(Error::SectionNotFound)
+        }
+    }
+
+    /// Write a section (serialize/translate) to a snapshot.
+    pub fn write_section<T>(&mut self, name: &str, object: &T) -> Result<usize>
+    where
+        T: Versionize,
+    {
+        let mut new_section = Section {
+            name: name.to_owned(),
+            data: vec![0; SNAPSHOT_MAX_SECTION_SIZE],
+        };
+
+        let slice = &mut new_section.data.as_mut_slice();
+        object.serialize(slice, &self.version_map, self.target_version)?;
+        // Resize vec to serialized section len.
+        let serialized_len =
+            slice.as_ptr() as usize - new_section.data.as_slice().as_ptr() as usize;
```

Comment:
This makes me slightly nervous, maybe `serialize` could return the number of bytes generated?